pages/developers/pytest.md

```diff
@@ -0,0 +1,254 @@
+
+* Simplicity: the easer / nicer you tests are to write, the more you will write.
+* Coverage: using as many inputs as possible increases the chances of finding something that breaks.
+* Performance: the faster the tests, the more situations you can run your tests in CI.
+* Reporting: when things break, you should get good information about what broke.
+
+> ### What about other choices?
+>
+> The alternative library, "nose", has been abandoned in favor of PyTest, which can run nose-style tests. The standard library has a test suite as well, but it's extremely verbose and complex; and since "developers" run tests, your test requirements don't affect users. And PyTest can run stdlib style testing too. So just use PyTest. All major packages use it too, including NumPy. Most other choices, like [Hypothesis][], are related to PyTest.
+
+### Basic test structure
+
+You should make a folder called "tests" in your repository (probably alongside your `src` or module folder, but rarely they are placed inside the module folder). Your files will be called `test_*.py`; PyTest's default discovery expects the word "test" to be in everything (yes, you can place tests alongside or inside your module because of this). This is an example of a test:
+
+`test_basic.py`:
+
+```python
+def test_funct():
+    assert 4 == 2**2
+```
+
+This looks simple, but it is doing several things:
+
+* The name of the function includes `test`, so it will run as a test.
+* The Python `assert` statement is rewritten by PyTest to capture exactly what happens. If it fails, you will get a clear, detailed report on what each value was.
+
+
+### Configuring PyTest
+
+PyTest supports configuration in `pytest.ini`, `setup.cfg`, or, since version 6, `pyproject.toml`. If you can require PyTest 6 (in other words, if Python 3.6+ is fine - PyTest is a developer requirement, not a user one, so limiting it is fine), then use `pyproject.toml`. This is an example configuration:
+
+```toml
+[tool.pytest.ini_options]
+minversion = "6.0"
+addopts = "-ra"
+testpaths = [
+    "tests",
+]
+```
+
+The `minversion` will print a nicer error if your PyTest is too old (though, ironically, it won't read this is the version is too old, so setting "6" or less in `pyproject.toml` is rather pointless). The addopts setting will add whatever you put there to the command line when you run; `-ra` will print a summary "r"eport of "a"ll results, which gives you a quick way to review what tests failed and were skipped, and why. Finally, `testpaths` will limit pytest to just looking in the folders given - useful if it tries to pick up things that are not tests from other directories. [See the docs](https://docs.pytest.org/en/stable/customize.html) for more options.
+
+PyTest also checks the current and parent directories for a `conftest.py` file. If it finds them, they will get run top-to-bottom. These files let you add fixtures and other python configuration per-directory.
```

Comment:
"and other python configuration per-directory." does not read well and is confusing.

Reply:
Hopefully it’s better now, let me know if it is not!

Reply:
Yep.